challenge_6/http_post.rb

```diff
@@ -0,0 +1,112 @@
+### Require dependencies
+require 'socket'
+require 'cgi'
+require 'uri'
+
+### Define some constants ###
+#############################
+# More info here: https://tools.ietf.org/html/rfc7231#section-6
+STATUS_CODES = {
+  ok: 200,
+  redirect: 303
+}
+
+# Accompanying text for status codes
+STATUS_CODES_TEXT = {
+  ok: 'OK',
+  redirect: 'See Other'
+}
+
+HTTP_VERSION = 'HTTP/1.1'
+
+BLOG_DATA = 
+[
+  {
+    title: 'My awesome blog!',
+    content: 'my favourite HTML tags are <p> and <script>'
+  },
+  {
+    title: 'Another cool blog!',
+    content: 'my favourite HTML tags are <br> and <hr>'
+  }
+]
+
+CRLF = "\r\n"
+#############################
+
+server = TCPServer.new 1234
+
+loop do
+  # Accept a client connection
+  client = server.accept
+  puts "Got a new client!"
+  
+  # Read the request line
+  request_line = client.readline.chomp
+  puts "Parsing HTTP request!"
+  method, target, http_version = request_line.split
+
+  puts "Building response for client!"
+  status_line = "#{HTTP_VERSION} #{STATUS_CODES[:ok]} #{STATUS_CODES_TEXT[:ok]}"
+  header_field = "Content-Type: text/html"
+
+  # Check request target to determine what to send back to client
+  if target == '/show-data'
+    message_body = ""
+    message_body << "<ul>"
+    BLOG_DATA.each do |element|
+      message_body << "<li>"
+      message_body << "<strong>Title: #{CGI.escape_html(element[:title])}</strong>, Content: #{CGI.escape_html(element[:content])}"
+      message_body << "</li>"
+    end
+    message_body << "</ul>"
+  elsif target == '/create-post' && method == 'POST'
+    status_line = "#{HTTP_VERSION} #{STATUS_CODES[:redirect]} #{STATUS_CODES_TEXT[:redirect]}"
+    # Don't need a Content-Type here, we're redirecting!
+    header_field = "Location: /show-data"
+
+    puts "Got a new POST request!"
+    # Not sure if this is how to do this
+    # Let's try reading all the headers in until we get a line that is just CRLF
+    headers = {}
+    line = client.readline
+    while (line = client.readline) != CRLF
+      header_name, header_value = line.chomp.split(": ")
+      headers[header_name] = header_value
+    end
+    content_length = headers["Content-Length"]
+    body = client.read(content_length.to_i)
+    post = Hash[URI.decode_www_form(body)]
+    post.transform_keys! { |key| key.to_sym }
```

Comment:
You could use the [`Symbol#to_proc` shorthand](https://rubydoc.info/stdlib/core/Symbol:to_proc) to write this as `post.transform_keys!(&:to_sym)` if you wanted to, although the block argument is more explicit.